FSDP namedtuple support

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83181
* #83150
* __->__ #83055
* #83035
* #82892

- NamedTuple support is blocking MultiModal adoption. TODO: add test

rohan-varma committed:
FSDP namedtuple support

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP namedtuple support"


- NamedTuple support is blocking MultiModal adoption. TODO: add test

[ghstack-poisoned]

rohan-varma committed:
Update on "FSDP namedtuple support"


- NamedTuple support is blocking MultiModal adoption. TODO: add test

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_utils.py:
```diff
@@ -3,6 +3,7 @@
 
 import torch
 from torch.nn.modules.batchnorm import _BatchNorm
+from torch.nn.parallel.scatter_gather import _is_namedtuple
```
Interesting that it comes from `scatter_gather`... 😆

awgu commented on test/distributed/fsdp/test_fsdp_misc.py:
```diff
@@ -52,6 +53,43 @@ def world_size(self):
+        for x in out:
+            self.assertNotEqual([], list(x._backward_hooks.values()))
+        # print(out._backward_hooks)
+        # res = torch.cat([e for e in out]).sum()
+        # res.backward()
+        # params = m.params
+        # for p in params:
+        #     full_param_padded = p._full_param_padded
+        #     print(full_param_padded.storage().size())
+        # Ensure they are resharded
+        dist.barrier()
```
1. Should the `res = ...` and the `res.backward()` be uncommented?
2. Is the `dist.barrier()` necessary?
3. Do we want to check that the `FlatParameter`(s) are resharded?

rohan-varma replied:
the test was failing when there was a backward() and no barrier(), I think it might have been a sync issue. 

Backward won't work here, due to the issues with the training state assert. Even if we fixed that, flat params won't be resharded. these issues are captured in: https://github.com/pytorch/pytorch/issues/83129, https://github.com/pytorch/pytorch/issues/83107 and will be follow up work.

awgu reviewed (approved):
Stamping to unblock, but I think the test needs a second look.

rohan-varma committed:
Update on "FSDP namedtuple support"


- NamedTuple support is blocking MultiModal adoption. TODO: add test

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge
